### ISSUE TITLE
benchmark macro fails in non-toplevel scopes

### ISSUE THREAD

**Issue by staticfloat** [johnmyleswhite/Benchmarks.jl#22]
Compare this:

``` julia
using Benchmarks
x = 10
@benchmark x*x
```

Which works, with this:

``` julia
using Benchmarks
let
  x = 10
  @benchmark x*x
end
```

Which outputs:

```
ERROR: UndefVarError: x not defined
```

I believe this is due to the attempted capture of local variables not working like it should.  @mbauman I am not a metaprogramming wizard.  Could you let me know if this is intended behavior or not?

**Comment by johnmyleswhite** [comment]
That's definitely what's happening. But @mbauman will have a better sense how we might fix it.

**Comment by staticfloat** [comment]
Note that my example above fails with a different error without [this fix](https://github.com/johnmyleswhite/Benchmarks.jl/pull/23).

**Comment by mbauman** [comment]
Ah, interesting.  I think the trouble is that I try to evaluate the type of `x` at parse time… and `x` isn't defined yet.  In fact, it has a really bizarre super-scope behavior, always evaluating the typeof x at the current module's top level.  Hm.

**Comment by mbauman** [comment]
I'd be fine simply disallowing this… except that I'm imagining that you have a super-useful use-case for this running benchmarks programmatically within some scoped block.  Which I really want to see happen :)

**Comment by staticfloat** [comment]
Well, it's nice to be able to `@benchmark` inside of a `for` loop.  Otherwise, I'll have to unroll all my loops.  Not the end of the world, but definitely not my favorite solution.

**Comment by mbauman** [comment]
Yeah, see my edit.

**Comment by staticfloat** [comment]
Why do we need to type-annotate the locals we're capturing?

**Comment by staticfloat** [comment]
If I just remove `nonconst_types`, and change `decls` to:

```
decls = Expr(:block, map((r,s)->:($r = $(esc(s))), nonconst_locals, nonconst)...)
```

(e.g. remove the type annotations) things "work".

**Comment by mbauman** [comment]
The idea here is to get benchmarking to behave exactly as though it were being used in real-life code.  And in real-life code, making things type-stable is step number one.  So we want to ensure that we're not timing the global dispatch machinery.

I think I have a solution.

**Comment by staticfloat** [comment]
Ah, I see, you don't want to include the method cache lookup that would happen if we use nonconcrete types.

**Comment by mbauman** [comment]
Fixed by #25
